source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp

```diff
@@ -0,0 +1,259 @@
+   */
+  virtual void clear();
+
+  /**
+   * @brief Get attribute list of trajectory times
+   */
+  const std::deque<std::chrono::nanoseconds>& get_times() const;
+
+  /**
+   * @brief Get attribute number of point in trajectory
+   */
+  int get_size() const;
+
+protected:
+  /**
+   * @brief Add new point and corresponding time to trajectory
+   */
+  template<typename DurationT>
+  void add_point(const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time);
+
+  /**
+   * @brief Insert new point and corresponding time to trajectory between two
+   * already existing points
+   */
+  template<typename DurationT>
+  void insert_point(const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time, int pos);
+
+  /**
+   * @brief Get the trajectory point at given index
+   * @param index the index
+   */
+  TrajectoryT& get_point(unsigned int index);
+
+  /**
+   * @brief Set the trajectory point at given index
+   * @param index the index
+   * @param point the new point
+   * @param new_time the new time
+   * @return Success of the operation
+   */
+  template<typename DurationT>
+  bool
+  set_point(unsigned int index, const TrajectoryT& point, const std::chrono::duration<int64_t, DurationT>& new_time);
+
+  /**
+   * @brief Set the trajectory point at given index
+   * @param points vector of new points
+   * @param new_time vector of new times
+   * @return Success of the operation
+   */
+  template<typename DurationT>
+  bool set_points(
+      const std::vector<TrajectoryT>& points, const std::vector<std::chrono::duration<int64_t, DurationT>>& new_times);
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   */
+  const std::pair<TrajectoryT, std::chrono::nanoseconds> operator[](unsigned int idx) const;
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   */
+  std::pair<TrajectoryT, std::chrono::nanoseconds> operator[](unsigned int idx);
+};
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase() : State() {
+  this->set_type(StateType::NONE);
+  this->reset();
+}
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase(const std::string& name) : State(name) {
+  this->set_type(StateType::NONE);
+  this->reset();
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::reset() {
+  this->State::reset();
+  this->points_.clear();
+  this->times_.clear();
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline void TrajectoryBase<TrajectoryT>::add_point(
+    const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time) {
+  this->set_empty(false);
+  this->points_.push_back(new_point);
+
+  if (!this->times_.empty()) {
+    auto const previous_time = this->times_.back();
+    this->times_.push_back(previous_time + new_time);
```

Comment:
I thought we were not doing that and the times that you provide are times from start, equivalent to JTC?

Reply:
well, we certainly store them as times from start, but the existing Trajectory functions were designed to just take a duration for that point, not time from start. So I kept the convention. 

But I would also prefer that we keep the same conventions across the stack (e.g., with JTC). I can make that change if we all agree

Reply:
hmm it's true that with the add and insert functions, maybe this requires more than I thought. Let me think about it